android/src/main/java/com/ahm/capacitor/camera/preview/CameraActivity.java

```diff
@@ -129,872 +189,1251 @@ public void setRect(int x, int y, int width, int height) {
-            Log.d(TAG, "Starting recording");
-            mRecorder.start();
-            eventListener.onStartRecordVideo();
-        } catch (IOException e) {
-            eventListener.onStartRecordVideoError(e.getMessage());
+        // Cannot find the one match the aspect ratio, ignore the requirement
+        if (optimalSize == null) {
+            minDiff = Double.MAX_VALUE;
+            for (Size size : sizes) {
+                if (Math.abs(size.getHeight() - targetHeight) < minDiff) {
+                    optimalSize = size;
+                    minDiff = Math.abs(size.getHeight() - targetHeight);
+                }
+            }
         }
+
+//        Log.d(TAG, "optimal preview size: w: " + optimalSize.getWidth() + " h: " + optimalSize.getHeight());
+        return optimalSize;
     }
 
-    public int calculateOrientationHint() {
-        DisplayMetrics dm = new DisplayMetrics();
-        Camera.CameraInfo info = new Camera.CameraInfo();
-        Camera.getCameraInfo(defaultCameraId, info);
-        int cameraRotationOffset = info.orientation;
-        Activity activity = getActivity();
+    private int getOrientationHint() {
+        int deviceRotation = getDeviceOrientation();
+        int sensorOrientation = getSensorOrientation();
+        return (sensorOrientation - deviceRotation + 360) % 360;
+    }
 
-        activity.getWindowManager().getDefaultDisplay().getMetrics(dm);
-        int currentScreenRotation = activity.getWindowManager().getDefaultDisplay().getRotation();
+    private int getDeviceOrientation() {
+        WindowManager windowManager = (WindowManager) activity.getSystemService(Context.WINDOW_SERVICE);
+        int rotation = windowManager.getDefaultDisplay().getRotation();
 
-        int degrees = 0;
-        switch (currentScreenRotation) {
+        int deviceRotation = 0;
+        switch (rotation) {
             case Surface.ROTATION_0:
-                degrees = 0;
+                deviceRotation = 0;
                 break;
             case Surface.ROTATION_90:
-                degrees = 90;
+                deviceRotation = 90;
                 break;
             case Surface.ROTATION_180:
-                degrees = 180;
+                deviceRotation = 180;
                 break;
             case Surface.ROTATION_270:
-                degrees = 270;
+                deviceRotation = 270;
                 break;
         }
+        return deviceRotation;
+    }
 
-        int orientation;
-        if (info.facing == Camera.CameraInfo.CAMERA_FACING_FRONT) {
-            orientation = (cameraRotationOffset + degrees) % 360;
-            if (degrees != 0) {
-                orientation = (360 - orientation) % 360;
-            }
+    private int getSensorOrientation() {
+        if (mCameraCharacteristics == null) return -1;
+        return mCameraCharacteristics.get(CameraCharacteristics.SENSOR_ORIENTATION);
+    }
+
+    private void configureTransform(int viewWidth, int viewHeight) {
```

Comment:
I still haven't pinned this one down, but something like:

```java
private void configureTransform(int viewWidth, int viewHeight) {
        if (cameraDevice == null || !textureView.isAvailable() || mPreviewSize == null) {
            return;
        }
        int rotation = activity.getWindowManager().getDefaultDisplay().getRotation();
        Matrix matrix = new Matrix();
        RectF viewRect = new RectF(0, 0, viewWidth, viewHeight);
        RectF bufferRect = new RectF(0, 0, mPreviewSize.getHeight(), mPreviewSize.getWidth());
        float centerX = viewRect.centerX();
        float centerY = viewRect.centerY();
        if (Surface.ROTATION_90 == rotation || Surface.ROTATION_270 == rotation) {
            bufferRect.offset(centerX - bufferRect.centerX(), centerY - bufferRect.centerY());
            matrix.setRectToRect(viewRect, bufferRect, Matrix.ScaleToFit.FILL);
            float scale = Math.max(
                    (float) viewHeight / mPreviewSize.getHeight(),
                    (float) viewWidth / mPreviewSize.getWidth());
            matrix.postScale(scale, scale, centerX, centerY);
            matrix.postRotate(90 * (rotation - 2), centerX, centerY);
        } else if (Surface.ROTATION_180 == rotation) {
            matrix.postRotate(180, centerX, centerY);
        }
        textureView.setTransform(matrix);
    }
```

Reply:
@eggbeard I have tested your PR and it works great.

Reply:
I have another fix to also include in this PR, so I'll try to make some time today to add this and @eggbeard's fix to the PR. Looks like I also need to resolve conflicts with master so will do that at the same time.

I suggest some exhaustive testing to find any other issues before merging in to master though.
I created this code for a specific use case (e.g. square preview) so have not extensively tested other generic usages.

Reply:
@dpa99c thank you very much for the prompt feedback. I will try to do more testing once you add your changes.

Reply:
having trouble getting this working for different preview aspect ratios so it's going to take me a bit longer to get it right